File: frontend/app/types/household.ts
import type { WeeklyMealPlan } from "~/types/meal-plan";
import type {
  AvailableIngredientInput,
  DietaryPreference,
  HealthPreference,
  NutritionTargetsInput,
  PricingMode,
} from "~/types/recommendation";

export interface HouseholdMemberInput {
  name: string;
  servings_per_meal: number;
  allergens: string[];
  excluded_ingredients: string[];
  dietary_preferences: DietaryPreference[];
}

export type PlannedMealType = "breakfast" | "lunch" | "dinner";
export type DishCourse = "main" | "side" | "salad" | "soup" | "breakfast" | "dessert" | "snack_appetizer" | "baked_good";

/** One dish position in a meal (ADR-0036). */
export interface MealRole {
  role_id: string;
  courses: DishCourse[];
  required: boolean;
}

/** Which meals of each day are planned, and each one's dishes (ADR-0046). */
export interface PlanShape {
  meals: Partial<Record<PlannedMealType, MealRole[]>>;
}

export interface HouseholdProfileInput {
  name: string;
  members: HouseholdMemberInput[];
  max_cooking_time_minutes: number;
  budget_per_meal_sgd: number | null;
  weekly_budget_sgd: number | null;
  health_preferences: HealthPreference[];
  nutrition_targets: NutritionTargetsInput;
  max_sodium_mg_per_meal: number | null;
  available_ingredients: AvailableIngredientInput[];
  pricing_mode: PricingMode;
  plan_shape?: PlanShape | null;
}

export interface HouseholdProfileUpdate extends HouseholdProfileInput {
  expected_version: number;
}

export interface HouseholdProfileVersion extends Omit<HouseholdProfileInput, "name"> {
  meal_composition?: MealRole[] | null;
  version: number;
  planning_household_size: number;
  allergens: string[];
  excluded_ingredients: string[];
  dietary_preferences: DietaryPreference[];
  created_at: string;
}

export interface HouseholdProfile {
  id: number;
  name: string;
  current_version: number;
  current: HouseholdProfileVersion;
  latest_plan_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface HouseholdPlanningOverrides {
  max_cooking_time_minutes?: number | null;
  budget_per_meal_sgd?: number | null;
  weekly_budget_sgd?: number | null;
  health_preferences?: HealthPreference[] | null;
  nutrition_targets?: NutritionTargetsInput | null;
  max_sodium_mg_per_meal?: number | null;
  available_ingredients?: AvailableIngredientInput[] | null;
  pricing_mode?: PricingMode | null;
}

export interface HouseholdProfilePlanRequest {
  start_date: string;
  profile_version?: number | null;
  overrides?: HouseholdPlanningOverrides;
}

export interface ProfileConstraintChange {
  field: string;
  before: unknown;
  after: unknown;
}

export interface HouseholdProfilePlanResult {
  profile_id: number;
  profile_version: number;
  replaces_plan_id: number | null;
  constraint_changes: ProfileConstraintChange[];
  plan: WeeklyMealPlan;
}

File: backend/app/schemas/household.py
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.meal_plan import WeeklyMealPlanResponse
from app.schemas.product import PricingMode
from app.schemas.recommendation import (
    AvailableIngredientInput,
    DietaryPreference,
    HealthPreference,
    NutritionTargets,
)


def _normalize_terms(values: list[str], *, ingredient_ids: bool = False) -> list[str]:
    normalized = {
        value.strip().lower().replace(" ", "_") if ingredient_ids else value.strip().lower()
        for value in values
        if value.strip()
    }
    return sorted(normalized)


class HouseholdMemberInput(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    servings_per_meal: int = Field(default=1, ge=1, le=3)
    allergens: list[str] = Field(default_factory=list, max_length=20)
    excluded_ingredients: list[str] = Field(default_factory=list, max_length=50)
    dietary_preferences: list[DietaryPreference] = Field(default_factory=list, max_length=4)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        return " ".join(value.split())

    @field_validator("allergens")
    @classmethod
    def normalize_allergens(cls, values: list[str]) -> list[str]:
        return _normalize_terms(values)

    @field_validator("excluded_ingredients")
    @classmethod
    def normalize_excluded_ingredients(cls, values: list[str]) -> list[str]:
        return _normalize_terms(values, ingredient_ids=True)


class HouseholdProfileWrite(BaseModel):
    name: str = Field(default="My household", min_length=1, max_length=120)
    members: list[HouseholdMemberInput] = Field(min_length=1, max_length=12)
    max_cooking_time_minutes: int = Field(default=60, ge=5, le=240)
    budget_per_meal_sgd: float | None = Field(default=None, gt=0, le=1000)
    weekly_budget_sgd: float | None = Field(default=None, gt=0, le=7000)
    health_preferences: list[HealthPreference] = Field(default_factory=list, max_length=3)
    nutrition_targets: NutritionTargets = Field(default_factory=NutritionTargets)
    max_sodium_mg_per_meal: float | None = Field(default=None, ge=100, le=5000)
    available_ingredients: list[AvailableIngredientInput] = Field(default_factory=list, max_length=100)
    pricing_mode: PricingMode = "fixture"

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        return " ".join(value.split())

    @model_validator(mode="after")
    def validate_total_servings(self) -> "HouseholdProfileWrite":
        if sum(member.servings_per_meal for member in self.members) > 12:
            raise ValueError("total servings_per_meal must not exceed 12")
        return self


class HouseholdProfileUpdate(HouseholdProfileWrite):
    expected_version: int = Field(ge=1)


class HouseholdProfileVersionResponse(BaseModel):
    version: int
    members: list[HouseholdMemberInput]
    planning_household_size: int
    max_cooking_time_minutes: int
    budget_per_meal_sgd: float | None
    weekly_budget_sgd: float | None
    allergens: list[str]
    excluded_ingredients: list[str]
    dietary_preferences: list[DietaryPreference]
    health_preferences: list[HealthPreference]
    nutrition_targets: NutritionTargets
    max_sodium_mg_per_meal: float | None
    available_ingredients: list[AvailableIngredientInput]
    pricing_mode: PricingMode
    created_at: datetime


class HouseholdProfileResponse(BaseModel):
    id: int
    name: str
    current_version: int
    current: HouseholdProfileVersionResponse
    latest_plan_id: int | None
    created_at: datetime
    updated_at: datetime


class HouseholdProfileVersionCollectionResponse(BaseModel):
    items: list[HouseholdProfileVersionResponse]


class HouseholdPlanningOverrides(BaseModel):
    max_cooking_time_minutes: int | None = Field(default=None, ge=5, le=240)
    budget_per_meal_sgd: float | None = Field(default=None, gt=0, le=1000)
    weekly_budget_sgd: float | None = Field(default=None, gt=0, le=7000)
    health_preferences: list[HealthPreference] | None = Field(default=None, max_length=3)
    nutrition_targets: NutritionTargets | None = None
    max_sodium_mg_per_meal: float | None = Field(default=None, ge=100, le=5000)
    available_ingredients: list[AvailableIngredientInput] | None = Field(default=None, max_length=100)
    pricing_mode: PricingMode | None = None


class HouseholdProfilePlanRequest(BaseModel):
    start_date: date = Field(default_factory=date.today)
    profile_version: int | None = Field(default=None, ge=1)
    overrides: HouseholdPlanningOverrides = Field(default_factory=HouseholdPlanningOverrides)


class ProfileConstraintChange(BaseModel):
    field: str
    before: Any
    after: Any


class HouseholdProfilePlanResponse(BaseModel):
    profile_id: int
    profile_version: int
    replaces_plan_id: int | None
    constraint_changes: list[ProfileConstraintChange]
    plan: WeeklyMealPlanResponse
